lib/src/metta/runner/pkg_mgmt/git_catalog.rs

```diff
@@ -0,0 +1,334 @@
+    /// Returns a new ModuleGitLocation.  This is a convenience; the usual interface involves deserializing this struct
+    pub(crate) fn new(url: String) -> Self {
+        let mut new_self = Self::default();
+        new_self.git_url = Some(url);
+        new_self
+    }
+    pub(crate) fn get_url(&self) -> Option<&str> {
+        self.git_url.as_ref().map(|s| s.as_str())
+    }
+}
+
+/// Struct that matches the catalog.json file fetched from the `_catalog.repo`
+#[derive(Serialize, Deserialize, Debug, Default)]
+struct CatalogFileFormat {
+    modules: BTreeMap<String, Vec<CatalogFileMod>>
+}
+
+impl CatalogFileFormat {
+    fn find_mods_with_name(&self, name: &str) -> Vec<ModuleDescriptor> {
+        let mut results = vec![];
+        if let Some(cat_mod_vec) = self.modules.get(name) {
+            for cat_mod in cat_mod_vec {
+                let uid = cat_mod.git_location.uid();
+                let descriptor = ModuleDescriptor::new(name.to_string(), cat_mod.version.clone(), Some(uid));
+                results.push(descriptor);
+            }
+        }
+        results
+    }
+    /// Scans the catalog looking for a single module that matches the provided descriptor
+    fn find_mod_with_descriptor(&self, descriptor: &ModuleDescriptor) -> Option<&CatalogFileMod> {
+        if let Some(cat_mod_vec) = self.modules.get(descriptor.name()) {
+            for cat_mod in cat_mod_vec.iter() {
+                if cat_mod.version.as_ref() == descriptor.version() {
+                    //NOTE ModuleDescriptors from GitCatalog always have a uid
+                    if Some(cat_mod.git_location.uid()) == descriptor.uid() {
+                        return Some(cat_mod);
+                    }
+                }
+            }
+        }
+        None
+    }
+    fn add(&mut self, new_mod: CatalogFileMod) -> Result<ModuleDescriptor, String> {
+        let uid = new_mod.git_location.uid();
+        let descriptor = ModuleDescriptor::new(new_mod.name.clone(), new_mod.version.clone(), Some(uid));
+        if self.find_mod_with_descriptor(&descriptor).is_none() {
+            let cat_mod_vec = self.modules.entry(new_mod.name.clone()).or_insert(vec![]);
+            cat_mod_vec.push(new_mod);
+        }
+        Ok(descriptor)
+    }
+}
+
+/// A single module in a catalog.json file
+#[derive(Clone, Debug, Serialize, Deserialize)]
+struct CatalogFileMod {
+    name: String,
+    #[serde(default)]
+    version: Option<semver::Version>,
+    #[serde(flatten)]
+    git_location: ModuleGitLocation,
+}
+
+impl CatalogFileMod {
+    fn new(name: String, version: Option<semver::Version>, git_location: ModuleGitLocation) -> Self {
+        Self {name, version, git_location}
+    }
+}
+
+/// Provides an interface to a git repo hosting a table of available modules
+#[derive(Debug)]
+pub struct GitCatalog {
+    name: String,
+    fmts: Arc<Vec<Box<dyn FsModuleFormat>>>,
+    refresh_time: u64,
+    catalog_repo: Option<CachedRepo>,
+    catalog_file_path: PathBuf,
```

Comment:
> What you describe is basically how it already works. The difference is mainly in the names.

I added this comment because `GitCatalog` contains `catalog_file_path` i.e. it manages the files on disk not only remote Git repository changes. Is it a piece of the code which can be removed by refactoring?

Reply:
Thanks Luke, after https://github.com/trueagi-io/hyperon-experimental/pull/691/commits/0fb5cc29a96ff6d90f63a17c48cbb1752fa36df6 it is more clear.